### ISSUE TITLE
Please support windows msys2

### ISSUE THREAD

**Issue by ahaoboy** [fish-shell/fish-shell#10956]
**YOUR TEXT HERE**

As a developer, I kindly request adding support for windows msys2, I understand that this is not part of the plan and might be extremely challenging, but as Windows developers, we desperately need a shell to replace pwsh. Currently, I use fish 3.7.x installed via msys2, and it works amazingly well—intelligent, fast, and efficient. If version 4.0 could also support windows msys2, it would be perfect.

![image](https://github.com/user-attachments/assets/ad1b4dc3-5f9f-4997-881b-a77ec45da917)

msys2-fish-3.7.1-3
https://packages.msys2.org/packages/fish?variant=x86_64

**Comment by faho** [comment]
That's not up to us.

Unfortunately, there is no cygwin/msys target for rust, and so no way to compile fish for it.

We had hoped it would appear in the last 2 years, but it appears the efforts have seized. See https://github.com/rust-lang/rust/issues/79854.

So there is nothing we can do.

**Comment by Aworldc** [comment]
Really? When I install msys2 and then normal windows rustup, it just works and I can use libraries installed in msys2, in rust, and compile that rust, without needing or using msvc or visual studio build tools, so I don't see why you'd need a special msys2 rust toolchain.

**Comment by faho** [comment]
Rustup will give you the *windows* target, which means everything that isn't available for windows, like `std::os::unix::fs::MetadataExt`, won't be available.

We specifically need a unix target, which would be a toolchain for msys/cygwin/whatever, targetting its unix features.

Fish is a unix shell that relies on unix features.

**Comment by faho** [comment]
If you try to compile for the x86_64-pc-windows-gnu target, you'll get some errors in our libc.c - which hooks up some libc constants that don't (yet) exist in the libc crate.

Stub those out and you get 3887 lines of errors, including:

- file descriptors ("unresolved import `std::os::fd`")
- file modes (mode_t/W_OK/X_OK/...)
- termios
- pgroups (tcsetpgrp/tcgetpgrp/setpgid)
- pid_t and uid_t
- `kill()`
- signals - SIGWINCH/SIGHUP/
- posix_spawn
- mmap
- readlink

Basically almost anything that talks to the OS would have to be changed. This touches 54 out of 185 .rs files.

That's not going to happen.

**Comment by SerJaimeLannister** [comment]
Have we tried https://github.com/uutils/coreutils/ , this is written in rust , so maybe this can help ?

**Comment by faho** [comment]
No, there will not be a windows port of fish.

https://github.com/fish-shell/fish-shell/issues/10956#issuecomment-2564671373 does not even reach the point of needing coreutils.

**Comment by SerJaimeLannister** [comment]
I totally understand this sentiment , Sorry if I sounded demanding.

**Comment by asukaminato0721** [comment]
someone did some patch, and the result:

https://github.com/Berrysoft/fish-shell/releases

hope it can be useful.

**Comment by faho** [comment]
> https://github.com/Berrysoft/fish-shell/releases

That involves a new rust target, btw: https://github.com/rust-lang/rust/pull/134999

Some of the patches look acceptable, some are a bit rough and need to be polished. Nothing we can't solve,
but we do require the rust target.

And it looks to be doing cross-compiling at first, I'm not sure if that's acceptable to the cygwin/msys maintainers.

**Comment by mati865** [comment]
https://github.com/Berrysoft/fish-shell/blob/27ebfb90f7e90d7cc9dc855e3d71a44bc09d9aa6/msys2/PKGBUILD this is MSYS2 PKGBUILD meaning it uses MSYS2 on Windows (yes, MSYS2 uses pacman and friends from Arch Linux) 😉

**Comment by lazka** [comment]
It's now in the main repo: https://packages.msys2.org/packages/fish
